### PR TITLE
Reuse previously resolved missing assembly metadata objects when identity doesn't match

### PR DESCRIPTION
**Pull request by tmat**
Fixes https://github.com/dotnet/try/issues/345.

**Background**
When the compiler is resolving compilation references and encounters a reference whose assembly identity does not match any of the explicitly given references it falls back to a Reference Resolver. In regular compilation an error is reported since there is no Reference Resolver. Interactive compiler supplies a resolver that attempts to find the missing reference. 

Once the missing reference is found the compiler adds it to the list of implicitly resolved assemblies and if the same missing assembly identity is referenced later (by a subsequent submission) it reuses the assembly symbol from this list, so that we don't end up loading multiple instances of the same missing assembly. [1]

The resolver may find an assembly that matches the name of the missing reference but doesn't match the version exactly. This is expected as a newer version of the assembly may be available than was referenced and the compiler correctly binds to the newer version. 

**Issues** 
The first submission added a reference to a netstandard20 library on .NET Core and the subsequent submission referenced a type from that library. The reference resolution in the compilation corresponding to the second submission ended up adding a new, distinct reference to the library, instead of reusing the one that was already resolved for the first submission compilation.

This was caused by mismatch in versions of some of the facades transitively referenced by the library thru `netstandard.dll` reference and resolved thru the missing reference resolution mechanism. The facades ware referenced with different versions than the actual versions available in the set of references the Reference Resolver resolved missing references from. The compiler ended up using two distinct instances of the same facade assembly metadata. An assembly symbol is only reused if all its transitively referenced assemblies match. In this case the library assembly symbol wasn't reused because some of the facade assembly symbols did not match.

The problem was that the mechanism [1] didn't reuse missing assembly metadata with a different version than what was requested. The fix is to reuse all successful resolutions based on referenced identity.

Another issue found during testing is handing resolution failures. We did not remember that a resolution of a given assembly identity failed and attempted to resolve the assembly again in the next submission. This may lead to inconsistencies and the same "can't convert T to T" errors since first submission might have a missing assembly symbol and next might create a real assembly symbol for the same assembly identity. The fix is to capture both successful and failed resolution results.

### CODE REVIEW

**Comment by tmat** [comment]
@AlekseyTs PTAL
/cc @rchande @colombod @jonsequitur @ManishJayaswal

**Comment by tmat** [comment]
@dotnet/roslyn-compiler PTAL

**Comment by AlekseyTs** [comment]
>         private readonly AssemblyMetadata _owner;

Although it might be convenient for lifetime management, it feels like we have created an unnecessary coupling between ```PEAssembly``` and ```AssemblyMetadata```. It looks like ```PEAssembly``` doesn't need anything from ```AssemblyMetadata``` for its functionality. Now we are taking dependency on this coupling not just for the purpose of lifetime management. This doesn't feel quite right.

---
Refers to: src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs:48 in fd58c35. [](commit_id = fd58c353a46ac7300d4355a938cf69e19308df75, deletion_comment = False)

File: src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs [hunk 1]
```diff
@@ -75,6 +75,8 @@ internal PEAssembly(AssemblyMetadata owner, ImmutableArray<PEModule> modules)
             _owner = owner;
         }
 
+        internal AssemblyMetadata Metadata => _owner;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs]: >internal AssemblyMetadata Metadata => _owner; [](start = 8, length = 45)

Do we really have to expose original object? It looks like we were not comfortable doing this in the past. Can we rely on ```GetNonDisposableMetadata``` API instead?

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs [hunk 2]
```diff
@@ -42,7 +42,7 @@ internal abstract class CommonReferenceManager
         internal abstract MetadataReference GetMetadataReference(IAssemblySymbol assemblySymbol);
         internal abstract ImmutableArray<MetadataReference> ExplicitReferences { get; }
         internal abstract ImmutableArray<MetadataReference> ImplicitReferences { get; }
-        internal abstract IEnumerable<KeyValuePair<AssemblyIdentity, PortableExecutableReference>> GetImplicitlyResolvedAssemblyReferences();
+        internal abstract IEnumerable<(PEAssembly, PortableExecutableReference)> GetImplicitlyResolvedAssemblyReferences();
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: >PEAssembly [](start = 39, length = 10)

Why do we need to store this explicitly? Can we get this from PortableExecutableReference?

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs [hunk 3]
```diff
@@ -677,11 +677,16 @@ public bool DeclarationsAccessibleWithoutAlias(int referencedAssemblyIndex)
...
         {
             foreach (PortableExecutableReference reference in ImplicitReferences)
             {
-                yield return KeyValuePairUtil.Create(ReferencedAssemblies[ReferencedAssembliesMap[reference]].Identity, reference);
+                var assembly = ReferencedAssemblies[ReferencedAssembliesMap[reference]];
+
+                // implicitly resolved references are PE assemblies
+                Debug.Assert(assembly.Assembly != null);
+
+                yield return (assembly.Assembly, reference);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: >assembly.Assembly [](start = 30, length = 17)

Is this where the new ```IAssemblySymbolInternal.Assembly``` API is used? Can we get the same information from PortableExecutableReference instead?

Comment (tmat) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: No, we can't. We need to reuse the AssemblyMetadata object that we already used before. PEReference may return a different instance if we read it again.

File: src/Compilers/Core/Portable/Symbols/IAssemblySymbolInternal.cs [hunk 4]
```diff
@@ -2,10 +2,17 @@
...
 namespace Microsoft.CodeAnalysis
 {
     internal interface IAssemblySymbolInternal : IAssemblySymbol
     {
         Version AssemblyVersionPattern { get; }
+
+        /// <summary>
+        /// Metadata of a PE assembly, null for other kinds of assembly symbols.
+        /// </summary>
+        PEAssembly? Assembly { get; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/IAssemblySymbolInternal.cs]: >PEAssembly? Assembly { get; } [](start = 8, length = 29)

It is not quite clear to me why we are adding this API. Let's discuss offline.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: > PEReference may return a different instance if we read it again.

I am assuming you are talking about a copy object. Why is it important to get the same instance? 

---
In reply to: [322395639](https://github.com/dotnet/roslyn/pull/38550#discussion_r322395639) [](ancestors = 322395639)

**Comment by AlekseyTs** [comment]
It feels to me that we are not fixing the the root cause of the issue. It looks like our goal is to address correctness problem and we are doing this by tuning symbol reuse infrastructure to make sure symbols are reused in more scenarios. While increasing the rate of symbols reuse is a good thing, it shouldn't be considered or used to address correctness issues. Symbol reuse is just an optimization, things should work the same way regardless whether some symbols are reused. No reuse of symbols should yield the same result as a full reuse.

**Comment by tmat** [comment]
This is addressing the core issue. We use the object identity of `AssemblyMetadata` to uniquely identity assembly metadata returned by a metadata reference. See https://github.com/dotnet/roslyn/blob/defeca5d0e44915fc978cfd93d3de2e4a6d0dbf4/src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs#L1015. 

The compiler should not request the assembly metadata object multiple times from the same metadata reference.

**Comment by AlekseyTs** [comment]
> The compiler should not request the assembly metadata object multiple times from the same metadata reference.

This can only improve symbol reuse, but symbol reuse is not a requirement for correctness. I believe the root cause is somewhere else. Improving symbol reuse is a good thing, but I do not think it actually fixes the underlying issue. Correctness shouldn't depend in symbol reuse.

**Comment by tmat** [comment]
`GetImplicitlyResolvedAssemblyReferences` returns the list of missing assemblies that were resolved in the previous submission. We use that list to avoid resolving these assembly identities again and we reuse the PEAssembly objects. Resolving the same assembly identity again would create a new assembly symbol for that identity.

**Comment by tmat** [comment]
We need to reuse exactly the same PEAssembly objects of the dependencies because otherwise the `IsMatchingAssembly` check would fail and we would create a new assembly symbol for a reference that references the dependency.

**Comment by AlekseyTs** [comment]
BTW, I think we can relax ```if (!ReferenceEquals(peAssembly.Assembly, Assembly)) ``` line to compare underlying MetadataIds. That should improve symbol reuse in the scenario without any additional APIs. However, I do not think that this would be a fix for the underlying issue that we are trying to fix in this PR. It also can lead to a worse symbol reuse on some other scenarios, because reuse for one assembly can lead to no reuse for several other assemblies.

**Comment by tmat** [comment]
@AlekseyTs Removed the reuse of PEAssebly

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs [hunk 5]
```diff
@@ -260,18 +260,15 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
...
-
-                        var data = ResolveMissingAssembly(binding.ReferenceIdentity, resolvedReference, importOptions, resolutionDiagnostics);
-                        if (data == null)
+                        if (!TryResolveMissingReference(
+                            requestingReference,
+                            binding.ReferenceIdentity,
+                            previouslyResolvedMissingReferencesOpt,
+                            resolver,
+                            resolutionDiagnostics,
+                            out var resolvedAssemblyIdentity,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: >var [](start = 32, length = 3)

Please spell out types for out vars.

File: src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs [hunk 6]
```diff
@@ -171,7 +171,7 @@ private PortableExecutableReference ResolveTrustedPlatformAssemblyCore(string na
 
             if (_lazyTrustedPlatformAssemblies.TryGetValue(name, out string path) && File.Exists(path))
             {
-                return MetadataReference.CreateFromFile(path, properties);
+                return _fileReferenceProvider(path, properties);
```
Comment (AlekseyTs) [inline src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: >return _fileReferenceProvider(path, properties); [](start = 16, length = 48)

Is this change necessary to fix the problem at hand?

File: src/Scripting/CSharpTest/InteractiveSessionReferencesTests.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            var libSource = @"
+public class C
+{
+    public readonly int X = 1;
+}
+";
+            var libImage = CreateCSharpCompilation(libSource, TargetFrameworkUtil.NetStandard20References, "lib").EmitToArray();
+
+            var dir = Temp.CreateDirectory();
```
Comment (AlekseyTs) [inline src/Scripting/CSharpTest/InteractiveSessionReferencesTests.cs]: >dir [](start = 16, length = 3)

Do we need to dispose this object in any way?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs (hunk 6 above)
Comment (tmat) [reply src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: No. But I kept it for consistency with the other functions that create metadata references in this class.

---
In reply to: [322469608](https://github.com/dotnet/roslyn/pull/38550#discussion_r322469608) [](ancestors = 322469608)

Comment (AlekseyTs) [reply src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: > No. But I kept it for consistency with the other functions that create metadata references in this class.

Please back this change by a unit-test that would fail without the change.

---
In reply to: [322495905](https://github.com/dotnet/roslyn/pull/38550#discussion_r322495905) [](ancestors = 322495905,322469608)

File: src/Scripting/CSharpTest/InteractiveSessionReferencesTests.cs (hunk 7 above)
Comment (tmat) [reply src/Scripting/CSharpTest/InteractiveSessionReferencesTests.cs]: Directories and files created off of `Temp` are automatically disposed by the base class.

**Comment by tmat** [comment]
@AlekseyTs Addressed the issues with failed resolutions we discussed yesterday and added more tests. PTAL

**Comment by tmat** [comment]
OK, I tried to make it easier to review by squashing away all the changes that I have undone. Will keep all commits this time.

**Comment by tmat** [comment]
The test failures are addressed now.

**Comment by tmat** [comment]
@cston For second review.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs [hunk 8]
```diff
@@ -464,37 +457,87 @@ private static void UpdateBindingsOfAssemblyBeingBuilt(ArrayBuilder<AssemblyRefe
...
+        /// Resolve <paramref name="referenceIdentity"/> using a given <paramref name="resolver"/>.
+        /// 
+        /// We make sure not to query the resolver for the same identity multiple times (across submissions).
+        /// Doing so ensures that we don't create multiple assembly symbols within the same chain of script compilations 
+        /// for the same implicitly resolved identity. Failure to do so results in cast errors like "can't convert T to T".
+        /// 
+        /// The method only records successful resolution results by updating <paramref name="implicitReferenceResolutions"/>.
+        /// Failures are only recorded after all resolution attempts have been completed.
+        /// 
+        /// This approach addresses teh following scenario. Consider a script:
```
Comment (cston) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: >teh [](start = 36, length = 3)

Typo.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs [hunk 9]
```diff
@@ -255,26 +245,22 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
...
-                            }
-                        }
-
-                        var data = ResolveMissingAssembly(binding.ReferenceIdentity, resolvedReference, importOptions, resolutionDiagnostics);
-                        if (data == null)
-                        {
-                            continue;
-                        }
+                        // one attempt for resolution succeeded (the attempt is cached in implicitReferenceResolutions, so further attempts won't fail and add it back):
+                        resolutionFailures.Remove(binding.ReferenceIdentity);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: >resolutionFailures.Remove(binding.ReferenceIdentity); [](start = 24, length = 53)

It would be good to add an explanation how could we first fail and then succeed for the same identity. We don't need to go back and fix something up for bindings it failed?

Comment (tmat) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: OK. The resolver takes 2 parameters: the metadata reference that references the identity and the identity. The default implementation of the resolver looks up the missing reference in the directory containing the metadata reference. Suppose the same identity is referenced from two metadata references from different directories and only one directory contains the assembly of the requested identity. Then one resolution will fail and the other will succeed. The failed resolutions are not "committed" until the end of the resolution, so we don't need to fix up anything.

Comment (tmat) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: The comment on TryResolveMissingReference explains. Maybe I should move it to the caller?

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs [hunk 10]
```diff
@@ -464,37 +457,87 @@ private static void UpdateBindingsOfAssemblyBeingBuilt(ArrayBuilder<AssemblyRefe
...
+        /// </code>
+        /// where both a.dll and b.dll reference x.dll, which is present only in dir2. Let's assume the resolver first 
+        /// attempts to resolve "x" referenced from "dir1\a.dll". The resolver may fail to find the dependency if it only
+        /// looks up the directory containing the referencing assembly (dir1). If we recorded and this failure immediately
+        /// we would not call the resolver to resolve "x" within the context of "dir2\b.dll" (or any other referencing assembly). 
+        /// 
+        /// This behavior would ensure consistency and if the types from x.dll do leak thru to the script compilation, but it 
+        /// would result in a missing assembly error. By recording the failure after all resolution attempts are complete
+        /// we also achieve a consistent behavior but are able to bind the reference to "x.dll". Besides, this approach
+        /// also avoids dependency on the order in which we evaluate the assembly references.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: >also avoids dependency on the order in which we evaluate the assembly references. [](start = 12, length = 81)

Is this true? Can different folders have different assemblies of the same identity? Then, the order of folders to look in is significant.

Comment (tmat) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: Right, in general, if there are multiple different versions of the assembly with the same identity then it indeed depends on the order. That usually doesn't happen for shipping libraries, but it can definitely happen for ad-hoc built libraries. The statement is made in the context of x.dll being only in one of the directories. Let me clarify that in the comment.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs (hunk 9 above)
Comment (tmat) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: Added clarification.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs (hunk 10 above)
Comment (tmat) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: Added clarification

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs [hunk 11]
```diff
@@ -255,26 +245,24 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
...
-                            }
-                        }
-
-                        var data = ResolveMissingAssembly(binding.ReferenceIdentity, resolvedReference, importOptions, resolutionDiagnostics);
-                        if (data == null)
-                        {
-                            continue;
-                        }
+                        // One attempt for resolution succeeded. The attempt is cached in implicitReferenceResolutions, so further attempts won't fail and add it back.
+                        // Since the failures tracked in resolutionFailures do not affect binding thre is no need to revert any decisions made so far.
```
Comment (cston) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs]: >thre [](start = 98, length = 4)

Typo.
